File: 318_maximum_product_of_word_lengths.java
class Solution {
    public int maxProduct(String[] words) {
        int len = words.length, maxProduct = 0;
        int[] nums = new int[len];
        for (int i = 0; i < len; i++) {
            for (char ch: words[i].toCharArray()) {
                // is using bit to represent if a char exists in the word. For example, if a and b exists, it will be 11 (binary).
                nums[i] = nums[i] | (1 << (ch -'a'));
            }
        }

        for (int i = 0; i < len; i++) {
            for (int j = 0; j < len; j++) {
                if ((nums[i] & nums[j]) == 0) {
                    maxProduct = Math.max(maxProduct, words[i].length() * words[j].length());
                }
            }
        }
        return maxProduct;
    }
}

File: python/318_maximum_product_of_word_lengths.py
class Solution:
    # use bit manipulation
    # adding the bit of a character into a single number
    # if nums[i] & nums[j] == 0 (no common words), we can use it as the candidate
    def maxProduct(self, words: List[str]) -> int:
        bit_nums = [0 for _ in range(len(words))]
        for i, word in enumerate(words):
            for c in word:
                bit_nums[i] = bit_nums[i] | (1 << ord(c) - ord('a'))

        longest = 0
        for i in range(len(words)):
            for j in range(i + 1, len(words)):
                if bit_nums[i] & bit_nums[j] == 0:
                    longest = max(longest, len(words[i]) * len(words[j]))
        return longest
